### ISSUE TITLE
aws s3 sync racing with local files

### ISSUE THREAD

**Issue by groner** [aws/aws-cli#2960]
I'm having an issue with S3 sync bombing when local files (which would be excluded anyway) are removed by another process.

I found issue #2333, and updated to the latest release, but the issue persists.  I think there is more that needs to be done to avoid these races.

**Comment by groner** [comment]
The test `test_sync_skips_over_files_deleted_between_listing_and_transfer` is testing for `get_file_stat` raising `OSError`, but `get_file_stat` actually translates `OSError` to `ValueError`.  I started down the path of making the test patch `os.stat` instead of `get_file_stat`, but it's a bit gnarly because there are unrelated calls to calls to `os.stat` that shouldn't be interfered with and then there are 6 calls to `os.stat` on the file used in the test, which means there may be several ways for a race to manifest here.

**Comment by kyleknap** [comment]
Interesting. Do you happen to have the debug logs on you when you run into this? You can get those by appending `--debug` to the command you ran. I'm specifically interested in the Traceback that you are getting. If you are unable to get those, is there at least a description coming back from the `ValueError`? The fix may be as easy as catching the `ValueError` here as well: https://github.com/aws/aws-cli/blob/1ae25c8773eb452a6d1a61d27ce8f99c47d99356/awscli/customizations/s3/filegenerator.py#L217

**Comment by groner** [comment]
I do happen to have the traceback handy still:
```
Traceback (most recent call last):
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/utils.py", line 211, in get_file_stat
    stats = os.stat(path)
FileNotFoundError: [Errno 2] No such file or directory: '/var/log/sterling/production.2017-11-14.b55df55ad9ce70642'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/s3handler.py", line 173, in call
    for fileinfo in fileinfos:
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/fileinfobuilder.py", line 31, in call
    for file_base in files:
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/comparator.py", line 79, in call
    src_file = advance_iterator(src_files)
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/filters.py", line 118, in call
    for file_info in file_infos:
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/filegenerator.py", line 142, in call
    for src_path, extra_information in file_iterator:
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/filegenerator.py", line 210, in list_files
    stats = self._safely_get_file_stats(file_path)
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/filegenerator.py", line 216, in _safely_get_file_stats
    size, last_update = get_file_stat(file_path)
  File "/usr/local/opt/awscli/lib/python3.5/site-packages/awscli/customizations/s3/utils.py", line 214, in get_file_stat
    path, e))
ValueError: Could not retrieve file stat of "/var/log/sterling/production.2017-11-14.b55df55ad9ce70642": [Errno 2] No such file or directory: '/var/log/sterling/production.2017-11-14.b55df55ad9ce70642'
```

I don't see that anything else calls `get_file_stat`, so the translation of OSError to ValueError could probably just be removed.

**Comment by jamesls** [comment]
Confirmed bug, thanks for reporting.

**Comment by matthiasr** [comment]
Proposed fix is in #3026.
